permutations/src/disjoint_cycles.rs

```diff
@@ -0,0 +1,261 @@
+    }
+
+    #[test]
+    pub fn parse_with_empty_string_returns_identity() {
+        let DisjointCycles(result) = "".parse::<DisjointCycles<4>>().unwrap();
+
+        let expected: Vec<Vec<usize>> = vec![];
+        assert_eq!(&result[..], &expected[..]);
+    }
+
+    #[test]
+    pub fn parse_with_missing_left_parenthesis_returns_error() {
+        let result = "1 2)".parse::<DisjointCycles<4>>();
+
+        // You could interpret this case as "values outside the parentheses"
+        // or an unmatched parenthesis.
+        assert!(
+            matches!(result, Err(PermutationError::UnmatchedParenthesis(_)))
+                || matches!(result, Err(PermutationError::ValueOutsideParentheses(_)))
+        );
+    }
+
+    #[test]
+    pub fn parse_with_missing_right_parenthesis_returns_error() {
+        let result = "(1 2".parse::<DisjointCycles<4>>();
+
+        assert!(matches!(
+            result,
+            Err(PermutationError::UnmatchedParenthesis(_))
+        ));
+    }
+
+    #[test]
+    pub fn parse_with_top_level_number_returns_error() {
+        let result = "1(2 3)".parse::<DisjointCycles<4>>();
+
+        assert!(matches!(
+            result,
+            Err(PermutationError::ValueOutsideParentheses(_))
+        ))
+    }
+
+    #[test]
+    pub fn parse_with_whitespace_between_cycles_is_handled_gracefully() {
+        let DisjointCycles(result) = "(0 3) (1 2)".parse::<DisjointCycles<4>>().unwrap();
+
+        let expected: Vec<Vec<usize>> = vec![vec![0, 3], vec![1, 2]];
+        assert_eq!(&result[..], &expected[..]);
+    }
+
+    #[test]
+    pub fn parse_with_whitespace_before_first_element_is_handled_gracefully() {
+        let DisjointCycles(result) = "( 1 2)".parse::<DisjointCycles<4>>().unwrap();
+
+        let expected: Vec<Vec<usize>> = vec![vec![1, 2]];
+        assert_eq!(&result[..], &expected[..]);
+    }
+
+    #[test]
+    pub fn parse_with_whitespace_after_last_element_is_handled_gracefully() {
+        let DisjointCycles(result) = "(1 2 )".parse::<DisjointCycles<4>>().unwrap();
+
+        let expected: Vec<Vec<usize>> = vec![vec![1, 2]];
+        assert_eq!(&result[..], &expected[..]);
+    }
+
+    #[test]
+    pub fn parse_with_nested_parentheses_returns_error() {
+        let result = "(1 (2 3) 4)".parse::<DisjointCycles<5>>();
+
+        assert!(matches!(result, Err(PermutationError::NestedCycle(_))))
+    }
+
+    #[test]
+    pub fn parse_with_bad_digit_returns_error() {
+        let result = "(1 2 a 3)".parse::<DisjointCycles<4>>();
+
+        assert!(matches!(result, Err(PermutationError::BadInt(_))))
+    }
+
+    #[test]
+    pub fn parse_with_single_cycle_parses_correctly() {
+        let DisjointCycles(cycles) = "(0 2 3)".parse::<DisjointCycles<4>>().unwrap();
+
+        let expected: Vec<Vec<usize>> = vec![vec![0, 2, 3]];
+        assert_eq!(&cycles[..], &expected[..])
+    }
+
+    #[test]
+    pub fn parse_with_multiple_cycles_parses_correctly() {
```

Comment:
more cases:

- a cycle is rewritten so the smallest element is first
- Multiple cycles are reordered so the smallest element is first
- challenge case that combines both of the above